src/main/java/com/google/devtools/build/lib/actions/ExecutionRequirements.java

```diff
@@ -105,6 +105,96 @@ public String parseIfMatches(String tag) throws ValidationException {
     }
   }
 
+  /** An execution requirement that is split into a key and a parseable value. */
+  @AutoValue
+  public abstract static class TaggedRequirement<ValueType> {
+    /**
+     * Thrown when a the {@link #validator()} method returns an error.
+     */
+    public static class ValidationException extends Exception {
+      private final String tagValue;
+
+      /**
+       * Creates a new {@link ValidationException}.
+       *
+       * @param tagValue the erroneous value that was parsed from the tag.
```

Comment:
nit: `@param` and `@return` take string fragments, which means they shouldn't have a trailing dot -- unless they contain more than one sentence. I know, it's confusing... More instances of this below.